File: project-euler/c/4.c
#include <stdio.h>

// Checks if a number is a palindrome.
int is_palindrome(int num)
{
	char str[6];
	sprintf(str, "%d", num);
	int j = 5;
	if (num < 100000) j = 4;
	for (int i = 0; i <= 2; i++)
	{
		if (str[i] != str[j]) return 0;
		j--;
	}
	return 1;
}

int main()
{
	int start = 100, largest = 0;
	do
	{
		for (int i = start; i <= 999; i++)
		{
			int result = start * i;
			if (is_palindrome(result) && result > largest) largest = result;
		}
		start++;
	} while(start < 1000);
	printf("%d", largest);
	return 0;
}

File: challenges/project-euler/python3/4.py
def is_palindrome(str):
    if len(str) < 2: return False
    start = str[:len(str)//2]
    end = str[len(str):len(str)//2 - 1:-1]
    if len(end) > len(start): end = end[:-1]
    return start == end

a, b, biggest = 999, 999, 0
while a >= 100 and b >= 100:
    test = a * b
    if (is_palindrome(str(test))):
        if test > biggest: biggest = test
    a -= 1
    if (a < 100):
        a = 999
        b -= 1

print(biggest)
